### DISCUSSION TITLE
Cable shutter direct IO connection

### DISCUSSION THREAD

**Discussion by LithiumNL** [frank26080115/alpha-fairy#18] [Q&A]
Out of curiosity.
Why do you connect the external shutter jack directly to an IO and not via a opto coupler or relay?
Connecting it directly could damage your input.

Its connected to pin 26 if I am correct, I think I'm looking wrong but I cannot find that pin in your software. Can you point me out where its declared?

**Comment by frank26080115** [comment]
By all means, add an optocoupler if you want. I just don't think it's necessary, and it'll drain the battery a bit faster. Thus I decided to make it easier to build, to be less intimidating to people who might not be as confident.

The pin assignment is defaulted to G26 https://github.com/frank26080115/alpha-fairy/blob/main/arduino_workspace/AlphaFairy/Settings.ino#L66

The actual toggling of the pin is done in `cam_shootQuickGpio` https://github.com/frank26080115/alpha-fairy/blob/main/arduino_workspace/AlphaFairy/CamUtils.ino#L25 , note that `SHUTTER_GPIO_ACTIVE_LOW` is the default, `SHUTTER_GPIO_ACTIVE_HIGH` is undefined (thus disabled), this is configured at compile-time (see `alfy_conf.h`) and cannot be accidentally changed during run-time. (as you can see, I have thought about this)

The enumeration to pin number translation is in `get_pinCfgGpio` https://github.com/frank26080115/alpha-fairy/blob/main/arduino_workspace/AlphaFairy/AppUtils.ino#L280

The fact that the popular Arduino Uno still runs at 5V, even in 2022, is partly to blame for so many people breaking things without optoisolation. The M5StickC-Plus is a whole different situation with a whole list of reasons that makes it much much safer. It uses lower IO voltage, there's easy way for the user to screw up electrically, and it doesn't execute bad code if the stack gets messed up. It's still possible (what if... what if... what if...) but it's so much safer that I decided to omit optoisolation.

**Reply by LithiumNL** [reply]
Frank, thanks for your quick answer!

Currently I am building a motorized camera slider with a EESP32 wroom32 chip. I came up your article and found it interesting if I can connect my Sony camera to the hotspot of the ESP but I found out that that would be a bit difficult and as you mentioned it could be unreliable.
I came across your page what the pinout of the jack is and did not expect to have it directly connected to the IO pin.

You say that the default is SHUTTER_GPIO_ACTIVE_LOW , this means that you switch between input and output. Is this the way you trigger the camera? Just out of curiosity :)
I would have in mind to make it a pull up and then make it just low when you want to make a picture.

**Reply by frank26080115** [reply]
> you switch between input and output

This is correct, it's called an "open drain" configuration. The resistor is inside the camera.

The code as it is today will work with an optocoupler without any changes. if you are building a motorized slider then you should add an optocoupler just in case of accidental generated back-EMF from the motor (ex, if somebody accidentally knocks the slider really hard, etc)

**Reply by LithiumNL** [reply]
Thats exactly why i was curios :) Any recommendations in what kind of octo coupler?

**Reply by frank26080115** [reply]
PC817 is a classic "jellybean" part for this sort of task

**Reply by LithiumNL** [reply]
Thanks!

Good luck with your project! Looks very promissing.
